File: src/code/go/03_stack/01_new_stack/array_stack.go
package new_stack

var _ Stack = (*ArrayStack)(nil)

type ArrayStack struct {
	data []any
	top  int
}

func NewArrayStack() *ArrayStack {
	return &ArrayStack{
		data: make([]any, 32),
		top:  -1,
	}
}

// Push 入栈
func (s *ArrayStack) Push(v any) {
	if s.top == cap(s.data)-1 {
		return
	}

	s.top += 1
	s.data[s.top] = v
}

// Pop 出栈
func (a *ArrayStack) Pop() any {
	if a.IsEmpty() {
		return nil
	}

	v := a.data[a.top]
	a.top--
	return v
}

// IsEmpty 是否为空
func (a *ArrayStack) IsEmpty() bool {
	return len(a.data) == 0 && a.top == -1
}

// Flush 清空
func (a *ArrayStack) Flush() {
	a.top = -1
}

// Top 置顶元素
func (a *ArrayStack) Top() any {
	if a.IsEmpty() {
		return nil
	}

	return a.data[a.top]
}

File: src/code/php/03_stack/01_new_stack/array_stack.php
<?php

require_once "./stack_interface.php";

/**
 * 数组栈
 */
class ArrayStack implements StackInterface
{
    /** array 存放数据 */
    private array $data;
    /** int 栈顶 */
    private int $top;

    public function __construct()
    {
        $this->top = -1;
    }

    /**
     * 添加元素
     *
     * @param mixed $val
     * @return void
     */
    public function push(mixed $val)
    {
        $this->top++;
        $this->data[$this->top] = $val;
    }

    /**
     * 弹出元素
     *
     * @return mixed
     */
    public function pop(): mixed
    {
        if ($this->isEmpty()) {
            return null;
        }

        $val = $this->data[$this->top];
        $this->top--;
        return $val;
    }

    /**
     * 判断是否为空
     *
     * @return boolean
     */
    public function isEmpty(): bool
    {
        return $this->top() == -1;
    }

    /**
     * 清空栈
     *
     * @return void
     */
    public function flush()
    {
        $this->top = -1;
    }

    /**
     * 返回栈顶元素
     *
     * @return mixed
     */
    public function top(): mixed
    {
        return $this->data[$this->top];
    }
}

$stack = new ArrayStack();
$stack->push(1);
$stack->push("a");
print($stack->top());
print($stack->pop());
print($stack->pop());
